lib/internal/child_process.js

```diff
@@ -384,12 +384,12 @@ ChildProcess.prototype.spawn = function(options) {
       continue;
     }
 
-    // The stream is already cloned and piped, thus close it.
+    // The stream is already cloned and piped, thus stop its readable side,
+    // otherwise we might attempt to read from the stream when at the same time
+    // the child process does.
     if (stream.type === 'wrap') {
-      stream.handle.close();
-      if (stream._stdio && stream._stdio instanceof EventEmitter) {
-        stream._stdio.emit('close');
-      }
+      stream.handle.readStop();
+      stream._stdio.pause();
```

Comment:
Instead of this, we could also use `stream._stdio.push(null)` and disable reading permanently, instead of until it’s explicitly started again. Any thoughts on that? Generally, the use case of “let another process read from this pipe, once that process is done we’ll continue reading ourselves” is not absurd or anything – shell scripts do that a lot.

Reply:
I never saw it myself but I can understand why just pausing the stream might be useful for piping logs or similar if the process crashes. If there's no technical drawback it sounds strictly better in term of enabled use cases, compared to closing it for good.

A few questions to better understand how it works:

- What does `readStop` do, particularly on writable streams? Is it just a noop?
- Maybe related, what does `pause` exactly do? Keep the fd open but stop polling it for data for readable streams, and buffer what's written into it for writable streams?
- Since you call both `pause` and `readStop`, how is it meant to be resumed? Is it just `resume()`, or is there another function that should be called to counteract `readStop`?

Reply:
> I never saw it myself but I can understand why just pausing the stream might be useful for piping logs or similar if the process crashes. If there's no technical drawback it sounds strictly better in term of enabled use cases, compared to closing it for good.

@arcanis Okay, thanks! I don’t think there are any technical drawbacks to this.

> A few questions to better understand how it works:
> 
> * What does `readStop` do, particularly on writable streams? Is it just a noop?

It stops libuv from listening for readable events on the fd until `readStart()` is called again. In partiuclar, no `read()` system call will occur during that time.

> * Maybe related, what does `pause` exactly do? Keep the fd open but stop polling it for data for readable streams, and buffer what's written into it for writable streams?

`pause()` makes the stream stop emitting `'data'` events. According to the [documentation](https://nodejs.org/api/stream.html#stream_readable_pause), it does not stop emitting `readable` events, but I guess in that case it’s up to the user to take care of not calling `.read()` anyway.

> * Since you call both `pause` and `readStop`, how is it meant to be resumed? Is it just `resume()`, or is there another function that should be called to counteract `readStop`?

Thanks for pointing this out – yes, one thing more is necessary here: Adding `stream.handle.reading = false;`, so that `._read()` will call `readStart()` again: https://github.com/nodejs/node/blob/eac8f50ca664186d099c03dcac71a22fde89671f/lib/net.js#L522, as well as resetting the stream state so that `_read()` actually gets called once we resume the stream. (None of these things are signs of a great streams implementation… we should probably get rid of `handle.reading` altogether, but that’s not really a concern for this PR.)

(Both calling `.resume()` explicitly and using `.on('data')` to start reading again should work.)

I’ve added those things, and also added a test file for this scenario.

Reply:
That's a lot of flags! 😄

And why isn't `readStop` enough? If the libuv doesn't listen for readable events anymore, I guess that the `data` events won't be triggered, right?

Reply:
@arcanis The issue that I’m seeing is, that at the time that this code is executed, there might be an existing `'data'` listener or similar that reads data from the stream. When the readable stream buffer drains, the streams implementation calls `_read()`, and for sockets that leads to a `handle.readStart()` call, undoing the effects of the `handle.readStop()` call here.

I mean, sure, it’s somewhat questionable to just ignore existing listeners that read data from the stream, but at this point I feel like that’s a reasonable solution?

Reply:
I see, that makes sense, thanks for the explanation!